internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,327 @@
+}
+
+func newTelemetryRecorder(
+	client awsxray.XRayClient,
+	sess *session.Session,
+	cfg *Config,
+	settings *awsutil.AWSSessionSettings,
+) Recorder {
+	recorder := &telemetryRecorder{
+		client:        client,
+		record:        newTelemetryRecord(),
+		recordUpdated: &atomic.Bool{},
+		interval:      defaultInterval,
+		done:          make(chan struct{}),
+		queue:         make(chan *xray.TelemetryRecord, queueSize),
+	}
+	if cfg != nil && cfg.IncludeMetadata {
+		if settings != nil {
+			recorder.resourceARN = settings.ResourceARN
+		}
+		recorder.hostname = os.Getenv(envAWSHostname)
+		recorder.instanceID = os.Getenv(envAWSInstanceID)
+
+		if settings == nil || !settings.LocalMode {
+			metadataClient := ec2metadata.New(sess)
+			if recorder.hostname == "" {
+				if hostname, err := metadataClient.GetMetadata("hostname"); err == nil {
+					recorder.hostname = hostname
+				}
+			}
+			if recorder.instanceID == "" {
+				if instanceID, err := metadataClient.GetMetadata("instance-id"); err == nil {
+					recorder.instanceID = instanceID
+				}
+			}
+		}
+	}
+	return recorder
+}
+
+func newTelemetryRecord() *xray.TelemetryRecord {
+	return &xray.TelemetryRecord{
+		SegmentsReceivedCount:  aws.Int64(0),
+		SegmentsRejectedCount:  aws.Int64(0),
+		SegmentsSentCount:      aws.Int64(0),
+		SegmentsSpilloverCount: aws.Int64(0),
+		BackendConnectionErrors: &xray.BackendConnectionErrors{
+			HTTPCode4XXCount:       aws.Int64(0),
+			HTTPCode5XXCount:       aws.Int64(0),
+			ConnectionRefusedCount: aws.Int64(0),
+			OtherCount:             aws.Int64(0),
+			TimeoutCount:           aws.Int64(0),
+			UnknownHostCount:       aws.Int64(0),
+		},
+	}
+}
+
+// Start starts the loop to send the records.
+func (t *telemetryRecorder) Start() {
+	t.startOnce.Do(func() {
+		go t.start()
+	})
+}
+
+// Stop closes the done channel to stop the loop.
+func (t *telemetryRecorder) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.done)
+	})
+}
+
+// start flushes the record once a minute if telemetry
+// data was recordUpdated.
+func (t *telemetryRecorder) start() {
+	ticker := time.NewTicker(t.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if t.recordUpdated.Load() {
+				t.add(t.cutoff())
+				t.recordUpdated.Store(false)
+				if failedToSend, err := t.send(t.flush()); err != nil {
+					for _, record := range failedToSend {
+						t.add(record)
+					}
+				}
+			}
+		case <-t.done:
+			return
+		}
+	}
+}
+
+// cutoff the current record and swap it out with a new record.
+// Sets the timestamp and returns the old record.
+func (t *telemetryRecorder) cutoff() *xray.TelemetryRecord {
+	t.mu.Lock()
+	oldRecord := t.record
+	t.record = newTelemetryRecord()
+	t.mu.Unlock()
+	oldRecord.SetTimestamp(time.Now())
+	return oldRecord
+}
+
+// add to the queue. If queue is full, drop the head of the queue and try again.
+func (t *telemetryRecorder) add(record *xray.TelemetryRecord) {
+	select {
+	case t.queue <- record:
+	default:
+		<-t.queue
```

Comment:
This means you're slightly dropping data, is there a way to capture this information else where?